podcasts/End of Year/Stories/ListeningTimeStory.swift

```diff
@@ -60,36 +72,98 @@ struct ListeningTimeStory: ShareableStory {
             StoryShareableText(L10n.eoyStoryListenedToShareText(listeningTime.storyTimeDescriptionForSharing))
         ]
     }
+}
 
-    private enum Constants {
-        /// The podcasts that are displayed on the view, the middle is your top 10 podcast
-        static let displayedPodcasts = [1, 0, 2]
-        static let coverSize = 180.0
+/// Given a GeometryProxy and a Double listening time it
+/// returns a view filled with circles representing days of listening.
+///
+/// The minimum amount of circles is 7.
+///
+/// The algorithm takes a `maxArea` and calculates how many balls it
+/// can fill there. However the returned `View` might be bigger or smaller,
+/// the area just give it some constraints to calculate values.
+struct CircleDays: View {
+    let proxy: GeometryProxy
+    let listeningTime: Double
 
-        static let spaceBetweenLabels = 22.0
-        static let labelHorizontalPadding = 35.0
+    var body: some View {
+        let maxArea = CGSize(width: proxy.size.width, height: proxy.size.height * 0.3)
+        let number = Double(listeningTime / 86400).rounded(.up)
+        let eachBallSquareArea = Double(maxArea.width * maxArea.height) / number
+        let numberOfBallsPerLine = max(7, (maxArea.width / eachBallSquareArea.squareRoot()).rounded(.up))
+        let numberOfLines = min((number / numberOfBallsPerLine).rounded(.up), max(1, (maxArea.height / eachBallSquareArea.squareRoot()).rounded(.up)))
+        let ballCalculatedWidth = maxArea.width / numberOfBallsPerLine
+        let ballCalculatedHeight = maxArea.height / numberOfLines
+        let ballPadding = min(ballCalculatedWidth, ballCalculatedHeight) * 0.05
+        let ballFinalWidth = min(ballCalculatedWidth, ballCalculatedHeight) - 4 * ballPadding
+
+        // The width of the days displayed as balls that the user didn't listened to podcasts
+        let missingDaysWidth = ((((numberOfLines * numberOfBallsPerLine) * 86400) - listeningTime) / 86400) * (ballFinalWidth + (2 * ballPadding))
+
+        // The content is repeated on the background so the gradient
+        // can have the exact same size as the circles.
+        VStack(spacing: 0) {
+            ForEach(0..<Int(numberOfLines), id: \.self) { _ in
+                HStack(spacing: 0) {
+                    ForEach(0..<Int(numberOfBallsPerLine), id: \.self) { _ in
+                        Circle()
+                            .foregroundStyle(.white)
+                            .frame(width: ballFinalWidth, height: ballFinalWidth)
+                            .padding(.all, ballPadding)
+                    }
+                }
+            }
```

Comment:
Could we just extract this into a function so we don't need to copy the code to use in the mask?